packages/lesswrong/lib/collections/posts/custom_fields.js

```diff
@@ -379,6 +379,88 @@ addFieldsDict(Posts, {
     hidden: false,
     control: "text"
   },
+  
+  // The next post. If a sequenceId is provided, that sequence must contain this
+  // post, and this returns the next post after this one in that sequence. If
+  // no sequenceId is provided, uses this post's canonical sequence.
+  nextPost: resolverOnlyField({
+    type: "Post",
+    graphQLtype: "Post",
+    viewableBy: ['guests'],
+    graphqlArguments: 'sequenceId: String',
+    resolver: async (post, { sequenceId }, { currentUser, Posts, Sequences }) => {
+      if (sequenceId) {
+        const nextPostID = await Sequences.getNextPostID(sequenceId, post._id);
+        if (nextPostID) {
+          const nextPost = await Posts.loader.load(nextPostID);
+          return accessFilterSingle(currentUser, Posts, nextPost);
+        }
+      }
+      if (post.canonicalNextPostSlug) {
+        const nextPost = await Posts.findOne({ slug: post.canonicalNextPostSlug });
+        return accessFilterSingle(currentUser, Posts, nextPost);
+      } else if(post.canonicalSequenceId) {
+        const nextPostID = await Sequences.getNextPostID(post.canonicalSequenceId, post._id);
+        if (nextPostID) return null;
+        const nextPost = await Posts.loader.load(nextPostID);
+        return accessFilterSingle(currentUser, Posts, nextPost);
+      }
+      
+      return null;
+    }
+  }),
+  
+  // The previous post. If a sequenceId is provided, that sequence must contain
+  // this post, and this returns the post before this one in that sequence.
+  // If no sequenceId is provided, uses this post's canonical sequence.
+  prevPost: resolverOnlyField({
+    type: "Post",
+    graphQLtype: "Post",
+    viewableBy: ['guests'],
+    graphqlArguments: 'sequenceId: String',
+    resolver: async (post, { sequenceId }, { currentUser, Posts, Sequences }) => {
```

Comment:
Really annoying that this shares 90% of the logic with the previous case. I think it's worth factoring this out into a single function.

Reply:
I actually don't think so. It's symmetrical, but, in order to refactor, the separated function would be taking field-names as part of its signature, which is terrible.